### DISCUSSION TITLE
Massive Abweichungen bei HTTP-Zählerstand Übernahme

### DISCUSSION THREAD

**Discussion by raldie** [camueller/SmartApplianceEnabler#533] [Q&A]
Hallo zusammen,

nachdem ich zum Test der kommenden Wärmepumpe unseren Herd als 3-phasiges Gerät eingebunden habe, ist mir mit Schrecken eine große Diskrepanz in den Werten der Zähler zum Sunny Portal aufgefallen. Ich habe seit ca. 1.5 Jahren den SAE mit 8 Tasmota Devices (Gosund SP1 und Sonoff POWR3) sehr erfolgreich und völlig problemlos im Einsatz. 58% Autarkie ohne einen Speicher sprechen schon für eine super Steuerungseffizienz mithilfe des SAE, die sonst nicht machbar wäre.

Zuerst habe ich meine Geräteanbindung überprüft, um meinen Fehler zu finden, durchsuchte alle vorhandene Logs's nach Error-Einträge, da die Stichproben bei den Leistungswerten korrekt waren. Alle Geräte sind per HTTP/Json/Zählerstand/5 Nachkommastellen/Tper60 angebunden und funktionieren augenscheinlich perfekt. Alle Tasmota-Geräte wurde für das jeweilige zu steuernde oder nur zu messende Gerät mit 0-Werten am Beginn angeschlossen und nicht mehr entfernt oder geändert. Updates wurden auch keine gemacht, die Tasmota v11 ist auch auf einem relativ neuen Stand.

Nachdem ich nichts gefunden habe, zog ich das Internet zurate und habe entsprechende Info's gefunden, dass das auch bei anderen Anwendern aufgetreten ist, jedoch keine Lösung herauslesen können.

![Sunny Portal 2](https://github.com/camueller/SmartApplianceEnabler/assets/101402200/3e538ff0-cc0c-4eaf-bb27-166e3b43783e)
![Spring](https://github.com/camueller/SmartApplianceEnabler/assets/101402200/760ed2e5-2374-45dc-8f42-b662300ee4a1)
![R600](https://github.com/camueller/SmartApplianceEnabler/assets/101402200/23dfcf60-301c-4402-a615-783f69a7d8f7)

![Tabelle](https://github.com/camueller/SmartApplianceEnabler/assets/101402200/832a6a21-16d8-40da-9dfb-c4ac3309cfa0)

Der Herd ist neu und der Trockner hat ein neues Device erhalten, da die Gosund defekt war.

Der EV-Charger hat eine Differenz von 1.5%, ich gehe davon aus, dass das WLAN-Probleme sind und man diese Differenz als normal ansehen kann, zumindest in unserer Anlage. Leider sind die über den SAE angebundenen Geräte mit sehr viel größeren Differenzen versehen, meiner Analyse nach um so mehr je kleiner die Leistung ist, Spitzenreiter ist hier der Rasenroboter.

Beim Studium bin auch über den Thread #427 gestolpert und hierdurch erfahren, dass es Axel ebenso mit seinen per HTTP angeschlossenen Geräten geht. Dort wie auch in anderen Diskussionen wurde auf die mangelnde Präzision der Tasmota Devices geschoben.
Bei genauerer Überlegung ist das bedauerlicherweise ein Trugschluss, da die unterschiedlichen Werte auf ungenaue Messwertaufnahmen geschoben wurden. Erst einmal klingt das logisch, jedoch liegt der Gedankenfehler in der Tatsache, dass ein falsch ermittelter Wert natürlich 1:1 im Portal wieder gespiegelt werden muss.

Wenn z.B. ein Gerät 150 Watt aufnimmt und der Tasmota Messwertaufnehmer nur 100 Watt ermittelt, dann wird auch das übertragen. Läuft dieses Gerät nun 24 Stunden, dann ist auf dem Tasmota WEB-Display ein Wert von 2.4 kWh für den Tag angezeigt. Das Sunny Portoal MUSS in diesem Fall für dieses Gerät natürlich auch nur 2.4 kWh anzeigen, da das Tasmota Device genau diesem Wert übermittelt hat. Das eigentlich 3.6 kWh (24 Std. × 0.15 kW) verbraucht wurden, ist nicht relevant!

Der Zählerstand dieses Tasmota Devices hat sind in den 24 Stunden um genau 2.4 kWh verändert, ergo muss der SAE auch diesen Wert im Laufe des ganzen Tages an das Portal per SEMP senden.

In #427 wurde der SDM72 der Wallbox als Überprüfung herangezogen, aber ich denke hier wurde nicht berücksichtigt, dass das eigentlich zwei verschiedene Prozesse sind, die Wallbox wurde per Modbus ausgelesen, die Tasmotas per HTTP. Zumindest die Messwertannahme und ggfls. auch Berechnung sind zwei unterschiedliche Routinen, die so nicht miteinander verglichen werden können.
Auffallend ist, dass die Leistung wohl eine enorme Auswirkung auf die Messwertermittlung hat, der Ladeziegel des Zweitwagens hat eine konstante Stromaufnahme und ist mit 2.3 kWh recht hoch, die Abweichung beträgt nur 2.31% und ist vollkommen in Ordnung. Der Rasenroboter hat mit seiner kleinen Leistungsaufnahme eine Abweichung von 205% !!!!!

Nur zur Erinnerung, alle Geräte wurden im SAE auf dieselbe Art und Weise angebunden, lediglich manche steuerbar und andere 'immer eingeschaltet'.

Da ich selbst seit knapp 40 Jahren Warenwirtschaftssysteme programmiere und kundenseitige Problemlösungen mein tägliches Brot sind, habe ich mir mal Gedanken über eine mögliche Fehlersuche gemacht. Da, wie oben beschrieben, die Genauigkeit des Geräts keinen Einfluss auf die Portalausgabe hat, wäre mein Lösungsansatz zur Fehlersuche eine unkomplizierte Möglichkeit der Überprüfung der Übertragungswerte. Allerdings ist die Addition aller SEMP-Zeilen eines Gerätes im Log aufwändig und daher nicht zielführend.

Die Tasmota's haben alle einen 'Gestern-Speicher', also den Wert, der an einem ganzen Tag verbraucht wurde.
Bei der SAE-Einstellung Zählerstand muss das ja dann auch der errechnete Wert aus den HTTP-Übernahmen eines Tages denselben Betrag ergeben.

Vorschlag an Axel:
Ich denke ein kleines eindimensionales ARRAY mit 255 Zellen addiert die jeweiligen HTTP-Geräte über die IP-Endnummer immer für einen Tag auf. Wenn dann um 0:00 Uhr das alte LOG abgeschlossen wird, brauchen nur alle Zellen <> 0 angehängt werden, z.B.: 192.168.2.171 = 5.44312 kWh

Evtl. auch als 1. Eintrag des neuen Logs wiederholen, dann ist es einfacher und schneller zu ersehen.
Je nachdem wo die Addition erfolgt, also in der HTTP-Übernahmeroutine oder in der SEMP-Erstellung könnte man die Problematik im SAE eingrenzen **oder natürlich auch ausschließen**. Vorteil diese Methode ist der, dass man jeden Tag vollkommen unabhängig die Werte überprüfen kann und man so einen Tag für die Kontrolle hätte, bevor die 'Gestern-Werte' im Device überschrieben werden.

Könnte man so etwas integrieren?
SAE-Testversionen werden sofort installiert!

LG
Ralf

**Comment by camueller** [comment]
Hast Du Deine Tasmota-Adapter auf MQTT umgestellt? Das könntest Du als erstes tun, damit es keine HTTP-Fehler mehr gibt.

Die an das Sunny Portal übertragenenen Werte kannst Du aus dem SAE-Log per grep/awk problemlos rausziehen und in eine Tabellenkalkulation importieren. Egal, wie gross oder klein die Abweichung ist: Was willst Du damit machen?

**Reply by raldie** [reply]
> Die an das Sunny Portal übertragenenen Werte kannst Du aus dem SAE-Log per grep/awk problemlos rausziehen und in eine Tabellenkalkulation importieren. Egal, wie gross oder klein die Abweichung ist: Was willst Du damit machen?

Ich wollte damit nichts machen, ich habe nur eine Möglich dargestellt und angeboten um das Problem der Differenz zwischen Messstelle und Sunny-Portal beim HTTP-Abruf des Zählerstandes einzukreisen. Klar, man kann die Daten per Excel aufaddieren, aber das ist zu aufwändig und macht keiner. Es geht doch darum heraus zu finden an welcher Stelle die kWh 'verschwinden'. Und der Vergleich einer Gesamtzahl eines Tages (gestern) im Log und der Tasmota-WEB-UI (auch gestern) ist schnell erledigt und bringt sofort Klarheit ob die Werte übereinstimmen oder wieviel sie differieren.

> Hast Du Deine Tasmota-Adapter auf MQTT umgestellt? Das könntest Du als erstes tun, damit es keine HTTP-Fehler mehr gibt.

Ich habe mich gestern damit mal beschäftigt und den Rasenroboter auf MQTT umgestellt, Zähler und Schalter.
Trozdem verstehe ich nicht, wie du darauf kommst, das es damit keine Fehler mehr gibt ?
Die Werte im MQTT-Json sind doch die selben wie bei dem HTTP-Json, der Zählerstand ist der selbe und die Abfragezeit ist bei mir auch jeweils 60 Sekunden, einmal durch den SAE (HTTP) und einmal durch das Tasmota-Device (tper -> mqtt)
Theoretisch sollten doch HTTP und MQTT beide das selbe Ergebnis im SAE zur Übergabe an das Portal erzeugen, oder ?

**Comment by camueller** [comment]
> Klar, man kann die Daten per Excel aufaddieren, aber das ist zu aufwändig und macht keiner.

Excel wird nur zur Summenbildung verwendet, den Rest machen grep/awk. Ist wesentlich einfacher, als wenn ich extra den Code anpassen muss.

>  Es geht doch darum heraus zu finden an welcher Stelle die kWh 'verschwinden'. 

Die Analyse hat ja schon gezeigt, dass es nur bei Verbrauchern mit geringer Leistungsaufnahme zu Abweichungen kommt. Demnach kann es sich eigentlich nicht um ein grundsätzliches Problem im SAE handeln.

> Ich habe mich gestern damit mal beschäftigt und den Rasenroboter auf MQTT umgestellt, Zähler und Schalter.
Trozdem verstehe ich nicht, wie du darauf kommst, das es damit keine Fehler mehr gibt ?

Manchmal kommt es bei der Abfrage von Tasmota-Adaptern zu HTTP-Fehlern. Abgesehen davon sollte HTTP und MQTT zum gleichen Ergebnis führen.

**Comment by raldie** [comment]
Wenn ich die Werte eines Tages ermitteln möchte, gehe ich dann davon aus, das ich die SEMP-Zeilen eines Gerätes mit dem Eintrag 'PowerInfo{averagePower=' verwende, hier die letzte Zeile des 0006-00 mit 7 Watt ?

_2023-06-12 14:50:56,105 DEBUG [http-nio-8080-exec-6] d.a.s.s.w.SempController [SempController.java:335] F-12343210-000000000006-00: Reporting device status from control
2023-06-12 14:50:56,106 DEBUG [http-nio-8080-exec-6] d.a.s.s.w.SempController [SempController.java:344] F-12343210-000000000006-00: DeviceStatus{eMSignalsAccepted=true, status=Off, errorCode=null}
2023-06-12 14:50:56,107 DEBUG [http-nio-8080-exec-6] d.a.s.s.w.SempController [SempController.java:350] F-12343210-000000000006-00: Reporting power info from meter.
2023-06-12 14:50:56,107 DEBUG [http-nio-8080-exec-6] d.a.s.s.w.SempController [SempController.java:364] F-12343210-000000000006-00: PowerInfo{averagePower=7, minPower=null, maxPower=null, timestamp=0, averagingInterval=60}_ 

Wie bitte habe ich dann die ermittelten Werte zu verarbeiten, einfach nur addieren oder in irgeneiner Form weiter zu berechnen?
Danke für eine kurze Info.

**Comment by camueller** [comment]
Mit dem nachfolgenden Befehl erzeugst Du eine CSV-Datei, welche für ein Gerät die in jeder Minute an den SHM gemeldete Leistung enthält (als Datum nicht den aktuellen Tag nehmen, sonst fehlen Werte und die Verzögeungen des Sunny Portal spielen eine Rolle):

```
grep F-xxxxxxxx-000000000007-00 /tmp/rolling-2023-06-06.log | grep PowerInfo | awk '{print $8}' | awk -F '=' '{print $2}' > /tmp/log.csv
```

Die erzeugte Datei sieht  so aus:
```
...
1,
1,
1,
2,
1,
41,
97,
94,
91,
89,
86,
84,
81,
79,
78,
75,
73,
71,
25,
...
```

Diese kannst Du in eine Tabellenkalulation laden, wobei jeder der 1464 Werte (hmmm - sollten eigentlich 24*60=1440 Werte sein) einer Wattminute entspricht.
In der zweiten Spalte teilst Du die Werte der ersten Spalte durch 60 und erhälst den Wert in Wattstunden.
Wenn Du die zweite Spalte aufsummierst, erhälst die die Wattstunden für den ganzen Tag. In meinem Fall sind das für meinen Mähroboter 85,91 Wh bzw. 0,08591 kWh. Im Sunny Portal sind für diesen Tag 0,043 kWh verzeichnet.

**Comment by raldie** [comment]
Danke für die gute Unterstützung, ich habe das mal ausprobiert, funktioniert, ist mir aber zur Überprüfung aller Geräte etwas aufwändig.
Ich habe da mal ein kleines .Net-Programm geschrieben und mir alle Werte aller SEMP-Geräte mit Typ oder Namen in einer Textdatei ausgeben lassen. Mit der Überprüfung von ca. 30 unterschiedlichen Tagen der Vergangenheit mit HTTP und/oder MQTT Anbindung stellte sich heraus, dass meine Anbindung an den SAE wie auch die SEMP-Übergabe des SAE's an den Home-Manager nur kleine Unterschiede beinhalteten und als Ursache für die Differenzen nicht infrage kommen.

Das Problem liegt eindeutig bei SMA. Da die Werte im Portal immer kleiner sind, gehen wahrscheinlich Datenpakete vom Home-Manager an den SMA-Webservice verloren. Dies erklärt aber leider nicht, warum die Differenzen bei Verbrauchern unter einem kW sehr viel höher sind. Eventuell ist es ja nach dem Umzug auf eine neue Serverlandschaft Ende des Monats besser.....

Da mein Analyseprogramm außer zur Kontrolle natürlich auch gut zum Test der Anbindung von Zählern an den SAE verwendbar ist, stelle ich es hier gerne für alle zur freien Verwendung ein. Einfach unter Windows starten, eine LOG-Datei per Drag&Drop auf das Tool ziehen und analysieren lassen. 

Hier noch ein Tipp für alle Windows-Anwender des SAE. Ich habe auf meinem Raspberry einfach Samba eingerichtet, SAE-Programmordner und SAE-Logordner freigegeben, und schon kann ich bequem von Windows aus darauf Zugriff nehmen, Backups durch Kopieren des Ordnerinhalts erstellen und auch SAE-Updates in Sekunden ausführen. Einfach mal nach 'Samba-Freigabe auf dem Raspberry Pi einrichten' googeln.

[SAE-SEMP-LOG-Analyse.zip](https://github.com/camueller/SmartApplianceEnabler/files/11744750/SAE-SEMP-LOG-Analyse.zip)

**Comment by camueller** [comment]
Ich gebe das mal an SMA weiter ...

**Comment by bollermann** [comment]
Hallo zusammen,
der Fehler tritt auch beim S0-Zähler meiner Wärmepumpe auf. Die im Sunnyportal angezeigte Energie ist deutlich niedriger als der aus den Zweirichtungszählern berechneten Energieaufnahme der Wärmepumpe. 

Der Fehler liegt zwischen 3% und 60% (Monatsbetrachtung), je nach Verbrauch der Wärmepumpe! Kleiner Verbrauch -> großer Fehler.

Das Portal lässt z.B. kleine Leistungen völlig unter den Tisch fallen. Wenn die Wärepumpe an einem Tag nicht lief, sondern nur die alte Steuerung+Raspi 11 W Dauerleistung zieht, zeigt das Portal als Tagesverbrauch 0 Wh an. Das ist auch beim Monatsverbrauch konsistent, das heißt der Monatsverbrauch ist exakt die Summe der im Portal angezeigten Tagesverbräuche. So werden bei mir (wenn die Wärmepumpe nicht laufen würde) im Monat ca. 7,9 kW unterschlagen. Ist leider nicht die ganze Wahrheit, da sich der Fehler bei mir damit nicht vollständig erklären lässt.

Kann ich davon ausgehen, dass der S0 Zähler (B+G E-Tech, 1000 Impulse/kWh) bei meiner Anwendung (11 W-3300 W) vernünftig funktioniert?

**Comment by camueller** [comment]
Habe noch kein Feedback zu dem Problem erhalten von SMA ...
